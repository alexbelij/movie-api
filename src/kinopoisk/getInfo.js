/* @flow */

import R from 'ramda';

import { imageUrlFromPath } from './utils';
import connector from './connector';
import getCredits from './getCredits';
import getStills from './getStills';

type KinopoiskApi$GetFilmResponse = {
  bigPosterURL?: ?string,
  country?: ?string,
  description?: ?string,
  filmID: string,
  filmLength?: ?string,
  genre?: ?string,
  imdbID?: ?string,
  nameEN?: ?string,
  nameRU: string,
  posterURL?: ?string,
  ratingAgeLimits?: ?string,
  ratingData: {
    rating?: ?string,
    ratingVoteCount?: ?string,
    ratingIMDb?: ?string,
    ratingIMDbVoteCount?: ?string,
    ratingFilmCritics?: ?string,
    ratingFilmCriticsVoteCount?: ?string,
  },
  ratingMPAA?: ?string,
  slogan?: ?string,
  type: 'KPFilm' | 'KPSerial',
  webURL: string,
  year?: ?string,
};

const parseRuntime = (rawRuntime: ?string): number => (
  rawRuntime
    ? R.pipe(
        R.split(':'),
        R.slice(0, 2),
        ([hours, minutes]: [string, string]): number => R.sum([
          parseInt(hours, 10) * 60,
          parseInt(minutes, 10),
        ]),
        (runtime: number) => (runtime === 0 ? NaN : runtime),
      )(rawRuntime)
    : NaN
);

const getInfo = async (id: number) => {
  const res: ?KinopoiskApi$GetFilmResponse =
    await connector.apiGet('getFilm', { filmID: id });

  if (!res) return null;

  const {
    description,
    filmID,
    filmLength,
    genre,
    imdbID,
    nameEN,
    nameRU,
    posterURL,
    ratingAgeLimits,
    ratingData = {},
    ratingMPAA,
    year,
  } = res;

  const credits = await getCredits(id);
  const stills = await getStills(id);

  return {
    kpId: parseInt(filmID, 10),
    imdbID,
    title: nameRU,
    originalTitle: nameEN,
    posterUrl: posterURL ? imageUrlFromPath(posterURL) : null,
    year: parseInt(year, 10),
    synopsis: description,
    runtime: parseRuntime(filmLength),
    genres: (genre || '').split(', '),
    ageRating: parseInt(ratingAgeLimits, 10),
    mpaaRating: ratingMPAA,
    kpRating: parseFloat(ratingData.rating),
    kpRatingVoteCount: parseInt(ratingData.ratingVoteCount, 10),
    imdbRating: parseFloat(ratingData.ratingIMDb),
    imdbRatingVoteCount: parseInt(ratingData.ratingIMDbVoteCount, 10),
    rtCriticsRating: parseInt(ratingData.ratingFilmCritics, 10),
    rtCriticsRatingVoteCount: parseInt(
      ratingData.ratingFilmCriticsVoteCount,
      10,
    ),
    stills,
    credits,
  };
};

export { parseRuntime as __parseRuntime };
export default getInfo;
