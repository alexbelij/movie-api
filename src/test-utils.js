/* @flow */

import R from 'ramda';

const childTypesForArray = R.memoize(
  R.pipe(R.map(R.type), R.uniq),
);

const modelFromObject = (obj: ?Object) => R.map(
  R.cond([
    [
      R.allPass([
        R.pipe(R.type, R.equals('Array')),
        R.pipe(childTypesForArray, R.length, R.equals(1)),
      ]),
      (val: Array<any>) => `Array<${R.pipe(childTypesForArray, R.head)(val)}>`,
    ],
    [R.pipe(R.type, R.contains(R.__, ['Object', 'Array'])), modelFromObject],
    [R.T, R.type],
  ]),
  obj,
);

const modelFromFirstElement = R.pipe(
  R.head,
  modelFromObject,
);
const modelFromFirstElementAtPath = (path: string) => R.pipe(
  R.pathOr([], path.split('.')),
  modelFromFirstElement,
);

export { modelFromObject, modelFromFirstElement, modelFromFirstElementAtPath };
