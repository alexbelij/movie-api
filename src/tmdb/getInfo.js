/* @flow */

import gql from 'graphql-tag';
import graphql from 'graphql-anywhere';
import R from 'ramda';

import { imageUrlFromPath } from './utils';
import connector from './connector';

type TmdbApi$Collection = {
  id: number,
  name: string,
  poster_path?: ?string,
  backdrop_path?: ?string,
};
type TmdbApi$Genre = {
  id: number,
  name: string,
};
type TmdbApi$ProductionCompany = {
  id: number,
  name: string,
};
type TmdbApi$ProductionCountry = {
  iso_3166_1: string,
  name: string,
};
type TmdbApi$Language = {
  iso_639_1: string,
  name: string,
};
type TmdbApi$CastMember = {
  cast_id: number,
  character: ?string,
  credit_id: string,
  id: number,
  name: string,
  order: number,
  profile_path: ?string,
};
type TmdbApi$CrewMember = {
  credit_id: string,
  department: string,
  id: number,
  job: string,
  name: string,
  profile_path: ?string,
};
type TmdbApi$Keyword = {
  id: number,
  name: string,
};
type TmdbApi$Video = {
  id: string,
  iso_639_1: string,
  iso_3166_1: string,
  key: string,
  name: string,
  site: 'YouTube',
  size: number,
  type: 'Trailer' | 'Teaser' | 'Clip' | 'Featurette',
};

type TmdbApi$GetMovieDetailsResponse = {
  adult: boolean,
  backdrop_path: ?string,
  belongs_to_collection: ?TmdbApi$Collection,
  budget: number,
  genres: Array<TmdbApi$Genre>,
  homepage: ?string,
  id: string,
  imdb_id: ?string,
  original_language: string,
  original_title: string,
  overview: string,
  popularity: number,
  poster_path: ?string,
  production_companies: Array<TmdbApi$ProductionCompany>,
  production_countries: Array<TmdbApi$ProductionCountry>,
  release_date: ?string,
  revenue: number,
  runtime: number,
  spoken_languages: ?Array<TmdbApi$Language>,
  status: 'Released',
  tagline: ?string,
  title: string,
  video: boolean,
  vote_average: number,
  vote_count: number,

  credits?: {
    cast: Array<TmdbApi$CastMember>,
    crew: Array<TmdbApi$CrewMember>,
  },
  keywords?: {
    keywords: Array<TmdbApi$Keyword>,
  },
  videos?: {
    results: Array<TmdbApi$Video>,
  },
};

const detailMethods = [
  'credits',
  'keywords',
  'videos',
];

type CastMember = {
  character: ?string,
  name: string,
  photoUrl: ?string,
};

const getCast = (
  credits: ?{ cast: Array<TmdbApi$CastMember> },
): Array<CastMember> => R.pipe(
  R.propOr([], 'cast'),
  R.map(({
    character,
    name,
    profile_path: profilePath,
  }: TmdbApi$CastMember): CastMember => ({
    character,
    name,
    photoUrl: profilePath ? imageUrlFromPath(profilePath, 300) : null,
  })),
)(credits);

type CrewMember = {
  job: string,
  name: string,
  photoUrl: string,
};

const getCreditsByJob = (
  credits: ?{ crew: Array<TmdbApi$CrewMember> },
  jobs: Array<string>,
): Array<CrewMember> => R.pipe(
  R.propOr([], 'crew'),
  R.filter(R.pipe(R.prop('job'), R.contains(R.__, jobs))),
  R.map(({ job, name, profile_path: profilePath }: TmdbApi$CrewMember) => ({
    job,
    name,
    photoUrl: profilePath ? imageUrlFromPath(profilePath, 300) : null,
  })),
)(credits);

const getInfo = async (id: number, query: void | string) => {
  const res: ?TmdbApi$GetMovieDetailsResponse = await connector.apiGet(
    `movie/${id}`,
    { append_to_response: detailMethods.join(',') },
  );

  if (!res) return null;

  const {
    backdrop_path: backdropPath,
    budget,
    genres,
    homepage,
    id: tmdbId,
    imdb_id: imdbId,
    original_language: originalLanguage,
    original_title: originalTitle,
    overview: synopsis,
    popularity: tmdbPopularity,
    poster_path: posterPath,
    production_companies: productionCompanies,
    production_countries: productionCountries,
    release_date: releaseDate,
    revenue,
    runtime,
    tagline,
    title,
    vote_average: tmdbRating,
    vote_count: tmdbRatingVoteCount,
    credits,
  } = res;

  const finalRes = {
    backdropUrl: backdropPath ? imageUrlFromPath(backdropPath, 1000) : null,
    budget,
    genres: genres.map(({ name }: TmdbApi$Genre) => name),
    homepage,
    tmdbId,
    imdbId,
    originalLanguage,
    originalTitle,
    synopsis,
    tmdbPopularity,
    posterUrl: posterPath ? imageUrlFromPath(posterPath, 500) : null,
    productionCompanies: productionCompanies.map(
      ({ name }: TmdbApi$ProductionCompany) => name,
    ),
    productionCountries,
    releaseDate,
    revenue: revenue === 0 ? null : revenue,
    runtime: runtime === 0 ? null : runtime,
    tagline,
    title,
    tmdbRating,
    tmdbRatingVoteCount,
    credits: {
      cast: getCast(credits),
      crew: {
        directors: getCreditsByJob(credits, ['Director']),
        producers: getCreditsByJob(credits, [
          'Associate Producer',
          'Co-Producer',
          'Executive Producer',
          'Producer',
        ]),
        writers: getCreditsByJob(credits, ['Screenplay', 'Characters']),
        composers: getCreditsByJob(credits, ['Original Music Composer']),
        cinematographers: getCreditsByJob(credits, ['Director of Photography']),
      },
    },
  };

  if (!query) return finalRes;

  return graphql(R.prop, gql`${query}`, finalRes);
};

export default getInfo;
