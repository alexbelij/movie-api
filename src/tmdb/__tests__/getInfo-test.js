/* @flow */

import { modelFromObject } from '../../test-utils';
import getInfo from '../getInfo';

const tmdbId = 140607; // Star Wars: The Force Awakens

describe('tmdb/getInfo', () => {
  it('fetches movie info from tmdb for a given id', async () => {
    expect(modelFromObject(await getInfo(tmdbId))).toMatchSnapshot();
  });

  it('formats response according to graphql query', async () => {
    const res = await getInfo(tmdbId, `
      {
        tmdbId
        imdbId
        title
        synopsis
      }
    `);

    expect(modelFromObject(res)).toMatchSnapshot();
  });
});
