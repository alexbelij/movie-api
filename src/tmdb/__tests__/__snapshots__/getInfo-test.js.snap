exports[`tmdb/getInfo fetches movie info from tmdb for a given id 1`] = `
Object {
  "backdropUrl": "String",
  "budget": "Number",
  "credits": Object {
    "cast": "Array<Object>",
    "crew": Object {
      "cinematographers": "Array<Object>",
      "composers": "Array<Object>",
      "directors": "Array<Object>",
      "producers": "Array<Object>",
      "writers": "Array<Object>",
    },
  },
  "genres": "Array<String>",
  "homepage": "String",
  "imdbId": "String",
  "originalLanguage": "String",
  "originalTitle": "String",
  "posterUrl": "String",
  "productionCompanies": "Array<String>",
  "productionCountries": "Array<Object>",
  "releaseDate": "String",
  "revenue": "Number",
  "runtime": "Number",
  "synopsis": "String",
  "tagline": "String",
  "title": "String",
  "tmdbId": "Number",
  "tmdbPopularity": "Number",
  "tmdbRating": "Number",
  "tmdbRatingVoteCount": "Number",
}
`;

exports[`tmdb/getInfo formats response according to graphql query 1`] = `
Object {
  "imdbId": "String",
  "synopsis": "String",
  "title": "String",
  "tmdbId": "Number",
}
`;
